File: apps/client/local_plugins/vendors/llm_openai/android/src/main/kotlin/com/aiagent/llm_openai/LlmOpenaiService.kt
package com.aiagent.llm_openai

import android.util.Log
import com.aiagent.plugin_interface.LlmCallback
import com.aiagent.plugin_interface.NativeLlmService
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import kotlin.coroutines.coroutineContext

/**
 * LlmOpenaiService — OpenAI-compatible LLM 原生服务实现
 *
 * - OkHttp SSE 流式推理
 * - 文本 delta：实时通过 onFirstToken / onTextDelta 派发
 * - tool_calls delta：按 index 累积 id/name/arguments，流结束后一次性派发
 *   onToolCallStart(id, name) + onToolCallArguments(完整 arguments) + onDone(fullText)
 * - 取消机制（call.cancel()）
 */
class LlmOpenaiService : NativeLlmService {

    companion object {
        private const val TAG = "LlmOpenaiService"
    }

    private val client = OkHttpClient()
    @Volatile private var activeCall: Call? = null

    private var apiKey: String = ""
    private var baseUrl: String = ""
    private var model: String = ""
    private var temperature: Double = 0.7
    private var maxTokens: Int = 2048
    private var systemPrompt: String? = null

    override fun initialize(configJson: String) {
        val cfg = JSONObject(configJson)
        apiKey = cfg.optString("apiKey", "")
        baseUrl = cfg.optString("baseUrl", "").trimEnd('/')
        model = cfg.optString("model", "")
        temperature = cfg.optDouble("temperature", 0.7)
        maxTokens = cfg.optInt("maxTokens", 2048)
        systemPrompt = cfg.optString("systemPrompt", "").ifBlank { null }
        Log.d(TAG, "initialize: baseUrl=$baseUrl model=$model")
    }

    override suspend fun chat(
        requestId: String,
        messages: List<Map<String, Any>>,
        tools: List<Map<String, Any>>,
        callback: LlmCallback,
    ): String {
        if (apiKey.isBlank() || baseUrl.isBlank() || model.isBlank()) {
            val errMsg = "LLM config incomplete: apiKey=${apiKey.isNotBlank()} baseUrl=${baseUrl.isNotBlank()} model=${model.isNotBlank()}"
            Log.e(TAG, errMsg)
            callback.onError("config_error", errMsg)
            return ""
        }

        // Build request body
        val messagesArray = JSONArray()
        if (!systemPrompt.isNullOrBlank()) {
            messagesArray.put(JSONObject().apply {
                put("role", "system")
                put("content", systemPrompt)
            })
        }
        for (msg in messages) {
            messagesArray.put(JSONObject(msg))
        }

        val body = JSONObject().apply {
            put("model", model)
            put("stream", true)
            put("messages", messagesArray)
            if (temperature != 0.7) put("temperature", temperature)
            if (maxTokens != 2048) put("max_tokens", maxTokens)
            if (tools.isNotEmpty()) {
                put("tools", JSONArray(tools.map { JSONObject(it) }))
            }
        }

        val url = if (baseUrl.endsWith("/chat/completions")) baseUrl
                  else "$baseUrl/chat/completions"

        val request = Request.Builder()
            .url(url)
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .header("Authorization", "Bearer $apiKey")
            .build()

        Log.d(TAG, "POST $url model=$model messages=${messages.size} tools=${tools.size}")

        val fullText = StringBuilder()
        var isFirstToken = true
        // tool_calls 累积器（按 OpenAI delta 的 index 字段）
        val toolBuilders = mutableMapOf<Int, _ToolCallBuilder>()

        return try {
            val response = suspendableEnqueue(request) ?: return ""

            if (!response.isSuccessful) {
                val respBody = response.body?.string()?.take(300) ?: ""
                val errMsg = "HTTP ${response.code}: $respBody"
                Log.e(TAG, "LLM request failed: $errMsg")
                callback.onError("http_${response.code}", errMsg)
                return ""
            }

            val source = response.body?.source() ?: return ""
            while (!source.exhausted()) {
                if (!coroutineContext.isActive) break

                val line = source.readUtf8Line() ?: break
                if (!line.startsWith("data: ")) continue
                val data = line.removePrefix("data: ").trim()
                if (data == "[DONE]") break

                val parsed = parseDelta(data) ?: continue

                // 文本增量
                parsed.contentDelta?.takeIf { it.isNotEmpty() }?.let { delta ->
                    fullText.append(delta)
                    if (isFirstToken) {
                        isFirstToken = false
                        callback.onFirstToken(delta)
                    } else {
                        callback.onTextDelta(delta)
                    }
                }

                // tool_calls 增量按 index 累积
                for (tc in parsed.toolCallDeltas) {
                    val builder = toolBuilders.getOrPut(tc.index) { _ToolCallBuilder() }
                    if (tc.id != null) builder.id = tc.id
                    if (tc.name != null) builder.name = tc.name
                    if (tc.argumentsDelta != null) builder.arguments.append(tc.argumentsDelta)
                }
            }

            // 一次性派发累积完成的 tool_calls
            for (b in toolBuilders.values) {
                val name = b.name ?: continue
                if (name.isBlank()) continue
                callback.onToolCallStart(b.id ?: "", name)
                callback.onToolCallArguments(b.arguments.toString())
            }

            callback.onDone(fullText.toString())
            fullText.toString()

        } catch (e: CancellationException) {
            activeCall?.cancel()
            ""
        } catch (e: IOException) {
            Log.e(TAG, "IO error: ${e.message}")
            callback.onError("io_error", e.message ?: "Unknown IO error")
            ""
        }
    }

    override fun cancel() {
        activeCall?.cancel()
    }

    // ─────────────────────────────────────────────────
    // 内部
    // ─────────────────────────────────────────────────

    private suspend fun suspendableEnqueue(request: Request): Response? {
        return suspendCancellableCoroutine { cont ->
            val call = client.newCall(request)
            activeCall = call
            cont.invokeOnCancellation { call.cancel() }
            call.enqueue(object : Callback {
                override fun onResponse(call: Call, response: Response) {
                    if (cont.isActive) cont.resumeWith(Result.success(response))
                }
                override fun onFailure(call: Call, e: IOException) {
                    if (cont.isActive) cont.resumeWith(Result.failure(e))
                }
            })
        }
    }

    private fun parseDelta(json: String): _DeltaParsed? {
        return try {
            val obj = JSONObject(json)
            val choices = obj.optJSONArray("choices") ?: return null
            if (choices.length() == 0) return null
            val delta = choices.getJSONObject(0).optJSONObject("delta")
                ?: return _DeltaParsed(null, emptyList())

            val content = delta.optString("content", "").ifBlank { null }

            val toolDeltas = mutableListOf<_ToolCallDelta>()
            delta.optJSONArray("tool_calls")?.let { tcs ->
                for (i in 0 until tcs.length()) {
                    val tc = tcs.optJSONObject(i) ?: continue
                    val fn = tc.optJSONObject("function")
                    toolDeltas.add(
                        _ToolCallDelta(
                            index = tc.optInt("index", 0),
                            id = tc.optString("id", "").ifBlank { null },
                            name = fn?.optString("name", "")?.ifBlank { null },
                            argumentsDelta = fn?.optString("arguments", "")?.ifBlank { null },
                        )
                    )
                }
            }
            _DeltaParsed(content, toolDeltas)
        } catch (_: Exception) {
            null
        }
    }

    private data class _DeltaParsed(
        val contentDelta: String?,
        val toolCallDeltas: List<_ToolCallDelta>,
    )

    private data class _ToolCallDelta(
        val index: Int,
        val id: String?,
        val name: String?,
        val argumentsDelta: String?,
    )

    private class _ToolCallBuilder {
        var id: String? = null
        var name: String? = null
        val arguments = StringBuilder()
    }
}

File: apps/client/local_plugins/vendors/llm_openai/ios/Classes/LlmOpenaiService.swift
import Foundation
import ai_plugin_interface
import os.log

/// OpenAI-compatible streaming chat completion (HTTP + SSE).
///
/// Mirrors the Android `LlmOpenaiService`:
///   - one in-flight request per service instance (newer cancels older).
///   - text deltas dispatched as they arrive (`firstToken` → `textDelta*`).
///   - `tool_calls` deltas accumulated by `index`; once the stream closes
///     the buffered tool calls are emitted as
///     `toolCallStart` + `toolCallArguments(complete JSON)` pairs.
///   - the function returns the full assistant text (also surfaced via
///     `onDone(fullText:)`).
public final class LlmOpenaiService: NativeLlmService {

    private static let log = OSLog(subsystem: "com.aiagent.llm_openai", category: "Service")

    private let session: URLSession
    private let stateLock = NSLock()
    private var activeTask: URLSessionDataTask?

    private var apiKey: String = ""
    private var baseUrl: String = ""
    private var model: String = ""
    private var temperature: Double = 0.7
    private var maxTokens: Int = 2048
    private var systemPrompt: String?

    public init() {
        let cfg = URLSessionConfiguration.ephemeral
        cfg.timeoutIntervalForRequest = 120
        cfg.timeoutIntervalForResource = 600
        // SSE streams produce small chunks frequently — disable any cellular
        // bandwidth heuristics that delay delivery.
        cfg.waitsForConnectivity = false
        self.session = URLSession(configuration: cfg)
    }

    public func initialize(configJson: String) {
        guard let data = configJson.data(using: .utf8),
              let cfg = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            os_log("initialize: invalid config json", log: Self.log, type: .error)
            return
        }
        apiKey = (cfg["apiKey"] as? String) ?? ""
        let rawBase = (cfg["baseUrl"] as? String) ?? ""
        baseUrl = stripTrailingSlash(rawBase)
        model = (cfg["model"] as? String) ?? ""
        if let t = cfg["temperature"] as? Double { temperature = t }
        else if let t = cfg["temperature"] as? Int { temperature = Double(t) }
        if let m = cfg["maxTokens"] as? Int { maxTokens = m }
        let sp = (cfg["systemPrompt"] as? String) ?? ""
        systemPrompt = sp.isEmpty ? nil : sp
        os_log("initialize: baseUrl=%{public}@ model=%{public}@",
               log: Self.log, type: .debug, baseUrl, model)
    }

    public func chat(
        requestId: String,
        messages: [[String: Any]],
        tools: [[String: Any]],
        callback: LlmCallback
    ) async throws -> String {
        guard !apiKey.isEmpty, !baseUrl.isEmpty, !model.isEmpty else {
            let msg = "LLM config incomplete: apiKey=\(!apiKey.isEmpty) baseUrl=\(!baseUrl.isEmpty) model=\(!model.isEmpty)"
            os_log("%{public}@", log: Self.log, type: .error, msg)
            callback.onError(code: "config_error", message: msg)
            return ""
        }

        // ── Build request ────────────────────────────────────────────
        var finalMessages: [[String: Any]] = []
        if let sp = systemPrompt, !sp.isEmpty {
            finalMessages.append(["role": "system", "content": sp])
        }
        finalMessages.append(contentsOf: messages)

        var body: [String: Any] = [
            "model": model,
            "stream": true,
            "messages": finalMessages,
        ]
        if abs(temperature - 0.7) > .ulpOfOne { body["temperature"] = temperature }
        if maxTokens != 2048 { body["max_tokens"] = maxTokens }
        if !tools.isEmpty { body["tools"] = tools }

        let url = baseUrl.hasSuffix("/chat/completions")
            ? baseUrl
            : "\(baseUrl)/chat/completions"
        guard let endpoint = URL(string: url) else {
            callback.onError(code: "config_error", message: "invalid baseUrl: \(url)")
            return ""
        }

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.httpBody = try JSONSerialization.data(withJSONObject: body, options: [])

        os_log("POST %{public}@ model=%{public}@ messages=%d tools=%d",
               log: Self.log, type: .debug, url, model, messages.count, tools.count)

        // ── Stream ────────────────────────────────────────────────────
        let stream = SseHttpStream(session: session, request: request)
        defer {
            stateLock.lock()
            if activeTask === stream.task { activeTask = nil }
            stateLock.unlock()
        }

        // Accumulators
        var fullText = ""
        var firstToken = true
        var toolBuilders: [Int: ToolCallBuilder] = [:]

        do {
            try await stream.start()
        } catch {
            callback.onError(code: "io_error", message: error.localizedDescription)
            return ""
        }
        stateLock.lock(); activeTask = stream.task; stateLock.unlock()

        // Validate HTTP status before consuming the body.
        if let http = stream.response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            let bodyPreview = await stream.drainBodyPreview(limit: 300)
            let msg = "HTTP \(http.statusCode): \(bodyPreview)"
            os_log("LLM request failed: %{public}@", log: Self.log, type: .error, msg)
            callback.onError(code: "http_\(http.statusCode)", message: msg)
            return ""
        }

        do {
            for try await line in stream.lines() {
                try Task.checkCancellation()
                guard line.hasPrefix("data: ") else { continue }
                let data = line.dropFirst(6).trimmingCharacters(in: .whitespacesAndNewlines)
                if data == "[DONE]" { break }

                guard let parsed = parseDelta(data) else { continue }

                if let delta = parsed.contentDelta, !delta.isEmpty {
                    fullText.append(delta)
                    if firstToken {
                        firstToken = false
                        callback.onFirstToken(textDelta: delta)
                    } else {
                        callback.onTextDelta(delta)
                    }
                }

                for tc in parsed.toolCallDeltas {
                    let builder = toolBuilders[tc.index] ?? ToolCallBuilder()
                    if let id = tc.id { builder.id = id }
                    if let name = tc.name { builder.name = name }
                    if let args = tc.argumentsDelta { builder.arguments.append(args) }
                    toolBuilders[tc.index] = builder
                }
            }
        } catch is CancellationError {
            stream.cancel()
            return ""
        } catch {
            os_log("IO error: %{public}@", log: Self.log, type: .error, error.localizedDescription)
            callback.onError(code: "io_error", message: error.localizedDescription)
            return ""
        }

        // Emit buffered tool calls (Android-side parity: one shot per call).
        for (_, builder) in toolBuilders {
            guard let name = builder.name, !name.isEmpty else { continue }
            callback.onToolCallStart(id: builder.id ?? "", name: name)
            callback.onToolCallArguments(builder.arguments)
        }

        callback.onDone(fullText: fullText)
        return fullText
    }

    public func cancel() {
        stateLock.lock()
        let task = activeTask
        activeTask = nil
        stateLock.unlock()
        task?.cancel()
    }

    // ── Parsing helpers ─────────────────────────────────────────────

    private struct DeltaParsed {
        let contentDelta: String?
        let toolCallDeltas: [ToolCallDelta]
    }

    private struct ToolCallDelta {
        let index: Int
        let id: String?
        let name: String?
        let argumentsDelta: String?
    }

    private final class ToolCallBuilder {
        var id: String?
        var name: String?
        var arguments: String = ""
    }

    private func parseDelta(_ jsonString: String) -> DeltaParsed? {
        guard let data = jsonString.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let choices = root["choices"] as? [[String: Any]],
              let first = choices.first else {
            return nil
        }
        let delta = (first["delta"] as? [String: Any]) ?? [:]
        let content: String? = {
            let s = (delta["content"] as? String) ?? ""
            return s.isEmpty ? nil : s
        }()

        var toolDeltas: [ToolCallDelta] = []
        if let tcs = delta["tool_calls"] as? [[String: Any]] {
            for tc in tcs {
                let fn = tc["function"] as? [String: Any]
                let id = (tc["id"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                let name = (fn?["name"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                let args = (fn?["arguments"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                toolDeltas.append(ToolCallDelta(
                    index: (tc["index"] as? Int) ?? 0,
                    id: id,
                    name: name,
                    argumentsDelta: args
                ))
            }
        }
        return DeltaParsed(contentDelta: content, toolCallDeltas: toolDeltas)
    }

    private func stripTrailingSlash(_ s: String) -> String {
        var out = s
        while out.hasSuffix("/") { out.removeLast() }
        return out
    }
}

